### PR TITLE
activation-level disillation

### LINKED ISSUE #385: Activation/feature-level distillation
# 🎯 **Goal (What & Why)**
Add activation-level distillation, usually leading to better student performance.

# 🚀 **Execution Plan**

### **Step 1: What is the smallest working version?**
- Distill based on all **mixer-layer outputs**
- Support the case where the student has the same number of layers as the teacher.
- Use MSE loss.
- Add a single coefficient that balances feature-level distillation vs logit-level.

As a first version:
$$L = L_{\text{logit}} + \lambda L_{\text{activation}}$$
with:
$$\mathcal{L}_{\text{activation}} = \frac{1}{N} \sum\_{i=1}^{N}||T_i(\mathbf{x}) - S_i(\mathbf{x})||_2$$
$T_i(x)$ and $S_i(x)$ denoting the outputs of the i-th layer's mixer of the teacher and student models. 

Details:
- teacher stores intermediate activations in `kwargs`
- student uses `kwargs` to compute activation-distillation losses, which it stores in `losses`

### **Step 2: What additional optimizations are possible (but optional)?**  
- Should support **TP with sequence-parallelism** (this is actually not optional, but can be done in a second step)
- Can configure which layers' outputs are used for distillation. For example, we could distill only based on mixer-layer outputs, or also based on MLP outputs, etc. Pass a {student -> teacher} mapping of layer-names to use for distillation
- Configurable loss: MSE, cosine, others?

# 📌 **Acceptance Criteria** (Must-Haves for Completion)
* The feature must be **functional and tested**.  
* The implementation must be **documented in practical terms**.  
* The PR must include a **performance/impact summary**.  
* **No refactors unless directly necessary** for feature completion.  

# 🛠️ **Project Management**
- [ ] **Assign the project to the Fast-LLM project.**
- [ ] **Set the `Estimate` field (in days) in the GitHub project.**
- [ ] **Use the `Size` field to categorize the PR size (Small/Medium/Large).**
- [ ] **Assign an owner when opening the issue.**

### PR DESCRIPTION
**Pull request by RaymondLi0**
# ✨ Description

Closes #385 

TODOs:

- [x] TP / sequence-tensor parallel: inconsistent gradients between TP=1 and TP=2
- [ ] Add tests: train with student==teacher, check that all losses are `0` and gradients as well.

Sanity checks:
- loading student and teacher from the same pretrained model gives `0` loss ✔️. But loss then increases to a small value instead of staying at 0. 
- Distilling from scratch with the same architecture doesn't lead to `0` loss (orange)
- Distilling the pretrained model, but with a sliding-window leads to low loss, lower with a larger window (green and purple)
Screenshot 2025-11-18 at 10 47 00 AM

<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01//EN" "http://www.w3.org/TR/html4/strict.dtd">
<html>
<head>
<meta http-equiv="Content-Style-Type" content="text/css">
<meta name="Generator" content="Cocoa HTML Writer">
<meta name="CocoaVersion" content="2575.7">
</head>
<body>

  | lm-loss | Logit distillation | Logit + Activation distillation
-- | -- | -- | --
Tokens/s/gpu | 3500 | 2900 | 2800
max_reserved (GB) | 44 | 77 | 78

</body>
</html>

With the caveat that distillation seems to experience memory spikes at specific points in training. The actual usage was lower most of the time:

Screenshot 2025-11-18 at 2 41 05 PM

## 🔍 Type of change

Select all that apply:

- [ ] 🐛 **Bug fix** (non-breaking change that addresses a specific issue)
- [x] 🚀 **New feature** (non-breaking change that adds functionality)
- [ ] ⚠️ **Breaking change** (a change that could affect existing functionality)
- [ ] 📈 **Performance improvement/optimization** (improves speed, memory usage, or efficiency)
- [ ] 🛠️ **Code refactor** (non-functional changes that improve code readability, structure, etc.)
- [ ] 📦 **Dependency bump** (updates dependencies, including Dockerfile or package changes)
- [ ] 📝 **Documentation change** (updates documentation, including new content or typo fixes)
- [ ] 🔧 **Infrastructure/Build change** (affects build process, CI/CD, or dependencies)

### Testing

- [ ] 🧪 I have added or updated tests to cover my changes.
- [ ] ✔️ New and existing tests pass locally with my changes.
- [ ] 🚦 I have tested these changes on GPUs and verified training stability.
- [ ] 🏋️ I have tested the changes on realistic training workloads, if applicable.

### Performance Impact

- [ ] 📊 I have run benchmarks where applicable to evaluate the performance impact.
- [ ] ✅ The benchmarks show no performance regression.
- [ ] 🚀 The benchmarks indicate a potential performance improvement.
- [ ] ⚠️ The benchmarks indicate a potential performance degradation.
- [ ] 📈 I have provided benchmark results and detailed any performance impact below, if applicable.

## 📊 Performance Impact Details

If there is any impact on performance, describe it and provide benchmark results, if applicable:

### CODE REVIEW

**Comment by tscholak** [comment]
great progress! did you freeze everything except the randomly initialized mixers?

File: fast_llm/engine/base_model/base_model.py [hunk 1]
```diff
@@ -179,6 +179,7 @@ def preprocess_batch(
         phase: PhaseType,
         iteration: int,
         metrics: dict | None = None,
+        setup_activation_storage: bool = False,
```
Comment (jlamypoirier) [inline fast_llm/engine/base_model/base_model.py]: Not needed, you can communicate through preprocessed kwargs.

File: tests/utils/model_configs.py [hunk 2]
```diff
@@ -538,6 +538,115 @@ def _update_and_add_testing_config(
...
 
+_update_and_add_testing_config(
+    # Tests logit distillation.
+    "mistral",
+    "mistral_distill_logits",
+    updates={
+        ("model", "base_model", "head", "distillation_model"): "teacher",
+        ("reference_models"): {
+            "teacher": {
+                "model": {
```
Comment (jlamypoirier) [inline tests/utils/model_configs.py]: You'll need complete model descriptions, ex. copied from another, otherwise the created model will be too big.

File: tests/utils/model_configs.py [hunk 3]
```diff
@@ -538,6 +538,115 @@ def _update_and_add_testing_config(
...
+    },
+    megatron_args=None,
+    checkpoint_format=MistralCheckpointFormat,
+    groups={
+        ModelTestingGroup.basic: ModelTestingGroupAction.normal,
+        ModelTestingGroup.checkpoint: ModelTestingGroupAction.unimportant,
+        ModelTestingGroup.convert: ModelTestingGroupAction.unimportant,
+        ModelTestingGroup.generate: ModelTestingGroupAction.unimportant,
+        ModelTestingGroup.megatron: ModelTestingGroupAction.not_implemented,
+        ModelTestingGroup.distributed: ModelTestingGroupAction.broken,  # failing: tp2, stp2, stp2_ce4
```
Comment (jlamypoirier) [inline tests/utils/model_configs.py]: We'll probably want to leave these as unimportant and run once in a while, because the testing suite can't really support many distributed runs.

File: fast_llm/layers/decoder/block.py [hunk 4]
```diff
@@ -148,6 +154,51 @@ def forward(
             hidden_states = torch.stack((fw_input, hidden_states), dim=0)
         return hidden_states
 
+    def activation_distillation_loss(self, hidden_states, bias, kwargs, losses):
+        """
+        Maybe apply activation distillation loss and setup backward hooks
+        """
+        mixer_output = hidden_states if bias is None else hidden_states + bias
```
Comment (jlamypoirier) [inline fast_llm/layers/decoder/block.py]: This should only be evaluated if needed.

File: tests/utils/model_configs.py (hunk 3 above)
Comment (RaymondLi0) [reply tests/utils/model_configs.py]: Makes sense. Some of the distributed tests are failing currently, so let's leave it as `broken` for now?

**Comment by RaymondLi0** [comment]
Thank you for the reviews! The comments are addressed, could you have another look? @jlamypoirier

File: fast_llm/layers/block/config.py [hunk 5]
```diff
@@ -164,3 +174,21 @@ def expanded_pattern(self) -> list[str]:
...
+            models.update(block.get_distillation_models())
+        return models
+
+    @classmethod
+    def _from_dict(cls, default: dict[str, typing.Any], strict: bool = True) -> typing.Self:
+        # Patch creeping type parameters from pretrained model
+        # TODO: fix this
+        if "block" in default:
+            removed = default.pop("block")
+            log(
```
Comment (jlamypoirier) [inline fast_llm/layers/block/config.py]: `logger.warning`?
